public/app/features/query/state/PanelQueryRunner.ts

```diff
@@ -194,10 +194,15 @@ export class PanelQueryRunner {
               }
             }
 
+            // replaceVariables() in the field config here so we end up diffing final values to increment structureRev
+            let customCfg = processedData.series[0].fields[1].config.custom;
+            customCfg.axisLabel = this.templateSrv.replace(customCfg.axisLabel, processedData?.request?.scopedVars);
+
```

Comment:
I am not following enough what the question is -- are you trying to find a way to expand variables generically, and not inside each transform?  

variables are can of worms -- the most notable issues is the they can represent a single value or an array.  When calling ".replace(" there is a third parameter that is used to describe how you want it expanded https://grafana.com/docs/grafana/latest/dashboards/variables/variable-syntax/#advanced-variable-format-options

Reply:
i want variables that are used in fieldconfig (like axisLabel) to properly work and be replaced in the config that we diff to increment structureRev and force a viz re-init.

try the attached dahsboard.

i think we can ignore vars that are arrays since we dont have any fieldConfigs that are arrays.

Reply:
The field / option config system already supports text interpolation, just do this: 

```patch
diff --git a/packages/grafana-ui/src/options/builder/axis.tsx b/packages/grafana-ui/src/options/builder/axis.tsx
index 70b2876bfb..2b8400be13 100644
--- a/packages/grafana-ui/src/options/builder/axis.tsx
+++ b/packages/grafana-ui/src/options/builder/axis.tsx
@@ -43,6 +43,7 @@ export function addAxisConfig(
       defaultValue: '',
       settings: {
         placeholder: 'Optional text',
+        expandTemplateVars: true,
       },
       showIf: (c) => c.axisPlacement !== AxisPlacement.Hidden,
       // Do not apply default settings to time and string fields which are used as x-axis fields in Time series and Bar chart panels
```

Reply:
it's a miracle! :tada: